### PR TITLE
[WebGL] Texture to tensor API

### PR DESCRIPTION
**Pull request by Linchenn**
This fixes https://github.com/tensorflow/tfjs/issues/3937

#### Changes
- Add a new method to the API: `tf.tensor(values: WebGLData, shape, dtype)` that allows users to create tensors from textures and `WebGLData` is an object of `WebGLTexture`, `height`, `width` and `channels` properties and the `channels` could be a non-empty subset of 'RGBA' in any order.

Reference:
https://github.com/tensorflow/tfjs/pull/6848 (The current PR starts from this PR of Na.)
https://github.com/tensorflow/tfjs/pull/4376

To see the logs from the Cloud Build CI, please join either our [discussion](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs) or [announcement](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs-announce) mailing list.

- - -
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/6853)

### CODE REVIEW

File: tfjs-backend-webgl/src/webgl.ts
Comment (qjia7) [inline tfjs-backend-webgl/src/webgl.ts:103]: Is it better to use the same name with `gl.texImage2D`, like `width/height`? So you can say `The width of the texture provided to gl.texImage2D during texture creation.`
`The height of the texture provided to gl.texImage2D during texture creation.`. Similar to the other params (internalFormat -> internalformat, textureFormat -> format, textureType -> type)?

File: tfjs-backend-webgl/src/webgl.ts
Comment (qjia7) [inline tfjs-backend-webgl/src/webgl.ts:135]: Since you require that the texel of input texture must be one value, it's not possible that `textureFormat` is equal to `gl.RGBA`?

In fact, I prefer you support both unpacked texture and the densely rgba texture as the input.

**Review by qjia7 — COMMENTED** [review]
The current design is to extend and expose the webgl backend's capacity directly but not expose it through tfjs-core. So you have come to an agreement to do it like this?
Two questions in my mind:
1. Usually, users check https://js.tensorflow.org/api/latest/ to get all latest APIs. How should we let users know the backend directly exposed APIs?
2. `dataToGPU` has been exposed by tfjs-core. It's wired that `createTensorFromTexture` is not like this way. And you also mentioned that the user can get texture by `tf.backend().getTexture(a.dataId)`. So `dataToGPU` is duplicated with `getTexture` from functionality?

**Review by lina128 — COMMENTED** [review]
Hi Lin, thank you for your hard work on this important feature! Sorry I probably wasn't clear when explaining the design of this method. I meant to say that let's follow Jiajia's Life of a Tensor design of the API, please refer to the doc for detail. Basically, the method should be in tensor.ts's constructor function. async tensor(...), it's at tfjs-core level. Also, we should support this two formats, R32F and RGBA, for R32, just pass through, for RGBA, do dense to unpacked using encodeMatrix. Thank you so much!!! I think only this two changes, plus one more test, then I'm all good. Also, can you also add the following reference to your PR description? Thanks. ref:https://github.com/tensorflow/tfjs/pull/6848

> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/6853#-:-NCs7qNr3ne7oRq3hCQP:b-z9iw37)* status: 0 of 1 approvals obtained (waiting on @Linchenn and @qjia7)

___
*[`tfjs-backend-webgl/src/backend_webgl.ts` line 181 at r1](https://reviewable.io/reviews/tensorflow/tfjs/6853#-NCs5_uM9GExKQmF7IGH:-NCs5_uM9GExKQmF7IGI:bvwgjah) ([raw file](https://github.com/tensorflow/tfjs/blob/1c119712673286c6aacf11e5201801fccf727837/tfjs-backend-webgl/src/backend_webgl.ts#L181)):*
> ```XML
> 
>   // Writes a new entry to the data store with a WebGL texture, and registers it
>   // to the texture manager.
> ```

The doccomment is outdated, update?



_Code quote:_
```XML
  // Writes a new entry to the data store with a WebGL texture, and registers it
  // to the texture manager.
```
___
*[`tfjs-backend-webgl/src/backend_webgl.ts` line 219 at r1](https://reviewable.io/reviews/tensorflow/tfjs/6853#-NCs6Evq7gPggh0AprFJ:-NCs6Evq7gPggh0AprFK:byenhz1) ([raw file](https://github.com/tensorflow/tfjs/blob/1c119712673286c6aacf11e5201801fccf727837/tfjs-backend-webgl/src/backend_webgl.ts#L219)):*
> ```XML
>     gpgpu_math.runProgram(
>         this.gpgpu, binary, inputTensorData, outputTensorData,
>         null /* customUniformValues */);
> ```

We can provide uniform values for texShape is useUniform flag is true, right?



_Code quote:_
```XML
/* customUniformValues */
```
___
*[`tfjs-backend-webgl/src/backend_webgl_test.ts` line 1331 at r1](https://reviewable.io/reviews/tensorflow/tfjs/6853#-NCs6VyOClacdbn6EYge:-NCs6VyOClacdbn6EYgf:b-lifrl) ([raw file](https://github.com/tensorflow/tfjs/blob/1c119712673286c6aacf11e5201801fccf727837/tfjs-backend-webgl/src/backend_webgl_test.ts#L1331)):*
> ```XML
>     const texture = gl.createTexture();
>     const tex2d = gl.TEXTURE_2D;
>     const internalFormat = gl.RGBA;
> ```

Can we add a test that takes a texture with RGBA format and densely laid out? Basically, for values 0-11, only need a texture with 3 texels.



_Code quote:_
```XML
gl.RGBA
```
___
*[`tfjs-backend-webgl/src/webgl.ts` line 97 at r8](https://reviewable.io/reviews/tensorflow/tfjs/6853#-NCgpR-PQ_F0Qnlerv18-r8-97:-NCs6t4lAksv8bxuyC1J:b-z8me46) ([raw file](https://github.com/tensorflow/tfjs/blob/c60a220aca1a08b3e9ca028188234fa540b76d30/tfjs-backend-webgl/src/webgl.ts#L97)):*
<details><summary><i>Previously, qjia7 (Jiajia Qin) wrote…</i></summary><blockquote>

Why we don't accept the densely rgba texture?
</blockquote></details>

Can we accept dense rgba texture? I think we want to support two texture formats: R32 and RGBA. And we check, if R32, then we just clone the texture, it's already unpacked, if RGBA, we run the encodeMatrix to convert the layout from dense to unpacked. Basically we output a tensor with unpacked texture (same behavior has creating a tensor with value on CPU, which will create an unpacked texture).
___
*[`tfjs-backend-webgl/src/webgl.ts` line 135 at r8](https://reviewable.io/reviews/tensorflow/tfjs/6853#-NCgpR-PQ_F0Qnlerv18-r8-135:-NCs7e6-0yu0xpEZAfZT:b675xnb) ([raw file](https://github.com/tensorflow/tfjs/blob/c60a220aca1a08b3e9ca028188234fa540b76d30/tfjs-backend-webgl/src/webgl.ts#L135)):*
<details><summary><i>Previously, qjia7 (Jiajia Qin) wrote…</i></summary><blockquote>

Since you require that the texel of input texture must be one value, it's not possible that `textureFormat` is equal to `gl.RGBA`?

In fact, I prefer you support both unpacked texture and the densely rgba texture as the input.
</blockquote></details>

Agreed with Jiajia.

**Comment by arpu** [comment]
this is amazing!  any code snipe for threejs ?

**Review by lina128 — APPROVED** [review]
The implementation is so clean, thank you Lin! LGTM. Clean up test and debug code before merge.

> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/6853#-:-NDU40Dh5o1rKysrLRLN:b-5snf8c)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @Linchenn and @qjia7)

File: tfjs-backend-webgl/src/webgl.ts
Comment (Linchenn) [reply tfjs-backend-webgl/src/webgl.ts]: done.

File: tfjs-backend-webgl/src/webgl.ts
Comment (Linchenn) [reply tfjs-backend-webgl/src/webgl.ts]: Done. Now, we support all non-empty subsequence of 'RGBA'.

**Review by Linchenn — COMMENTED** [review]
Thank you Na and Jiajia! Now, 'RGBA', 'GA', 'R', etc channels are all supported. Also, if the size of texture is not matched with the size of tensor, we will pad zeros or ignore rear values in the texture.

> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/6853#-:-NDURPoU3j90doAfOmPF:bha49uj)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @lina128, @Linchenn, and @qjia7)

___
*[`tfjs-backend-webgl/src/backend_webgl.ts` line 181 at r1](https://reviewable.io/reviews/tensorflow/tfjs/6853#-NCs5_uM9GExKQmF7IGH:-NDUR9kt3V91Uvw-QfwP:b-7rl0jd) ([raw file](https://github.com/tensorflow/tfjs/blob/1c119712673286c6aacf11e5201801fccf727837/tfjs-backend-webgl/src/backend_webgl.ts#L181)):*
<details><summary><i>Previously, lina128 (Na Li) wrote…</i></summary><blockquote>

The doccomment is outdated, update?
</blockquote></details>

done.
___
*[`tfjs-backend-webgl/src/backend_webgl.ts` line 219 at r1](https://reviewable.io/reviews/tensorflow/tfjs/6853#-NCs6Evq7gPggh0AprFJ:-NDURFQRAQvSWEAayHbU:b-896fix) ([raw file](https://github.com/tensorflow/tfjs/blob/1c119712673286c6aacf11e5201801fccf727837/tfjs-backend-webgl/src/backend_webgl.ts#L219)):*
<details><summary><i>Previously, lina128 (Na Li) wrote…</i></summary><blockquote>

We can provide uniform values for texShape is useUniform flag is true, right?
</blockquote></details>

Done.
___
*[`tfjs-backend-webgl/src/backend_webgl_test.ts` line 1331 at r1](https://reviewable.io/reviews/tensorflow/tfjs/6853#-NCs6VyOClacdbn6EYge:-NDURIgC0nGYxfZdPVf0:b-896fix) ([raw file](https://github.com/tensorflow/tfjs/blob/1c119712673286c6aacf11e5201801fccf727837/tfjs-backend-webgl/src/backend_webgl_test.ts#L1331)):*
<details><summary><i>Previously, lina128 (Na Li) wrote…</i></summary><blockquote>

Can we add a test that takes a texture with RGBA format and densely laid out? Basically, for values 0-11, only need a texture with 3 texels.
</blockquote></details>

Done.
___
*[`tfjs-backend-webgl/src/webgl.ts` line 97 at r8](https://reviewable.io/reviews/tensorflow/tfjs/6853#-NCgpR-PQ_F0Qnlerv18-r8-97:-NDURKRQBh7UB8K1HZqt:b-896fix) ([raw file](https://github.com/tensorflow/tfjs/blob/c60a220aca1a08b3e9ca028188234fa540b76d30/tfjs-backend-webgl/src/webgl.ts#L97)):*
<details><summary><i>Previously, lina128 (Na Li) wrote…</i></summary><blockquote>

Can we accept dense rgba texture? I think we want to support two texture formats: R32 and RGBA. And we check, if R32, then we just clone the texture, it's already unpacked, if RGBA, we run the encodeMatrix to convert the layout from dense to unpacked. Basically we output a tensor with unpacked texture (same behavior has creating a tensor with value on CPU, which will create an unpacked texture).
</blockquote></details>

Done.

**Comment by Linchenn** [comment]
I have tested also:
1. texture is f16, tfjs is f32 (no precision loss)
2. texture is f32, tfjs is f16 (can capture numbers in f16 range)

The both worked as expected. Since these unit tests' codes are too tricky and this case is rare, I did not add them.

**Review by lina128 — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/6853)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @lina128, @Linchenn, @pyu10055, and @qjia7)

___
*[`tfjs-backend-webgl/src/backend_webgl_test.ts` line 1432 at r23](https://reviewable.io/reviews/tensorflow/tfjs/6853#-NDUgfC_6G892FhGjgT0:-NDUgfC_6G892FhGjgT1:b-3mwtyj) ([raw file](https://github.com/tensorflow/tfjs/blob/2f66265e34053eab94b0b1635f892d56e01e76c6/tfjs-backend-webgl/src/backend_webgl_test.ts#L1432)):*
> ```XML
>   it('works for f16', async () => {
>     // Unlike in the basic usage test, rather than creating a texture from
>     // cratch, we must extract the output texture from an operation because we
> ```

scratch



_Code quote:_
```XML
cratch
```

**Review by Linchenn — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/6853)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @lina128, @Linchenn, @pyu10055, and @qjia7)

___
*[`tfjs-backend-webgl/src/backend_webgl_test.ts` line 1432 at r23](https://reviewable.io/reviews/tensorflow/tfjs/6853#-NDUgfC_6G892FhGjgT0:-NDUhNme8RhrmK44i1GU:boyexb3) ([raw file](https://github.com/tensorflow/tfjs/blob/2f66265e34053eab94b0b1635f892d56e01e76c6/tfjs-backend-webgl/src/backend_webgl_test.ts#L1432)):*
<details><summary><i>Previously, lina128 (Na Li) wrote…</i></summary><blockquote>

scratch
</blockquote></details>

Done. Thanks!

**Review by pyu10055 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/6853)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @lina128, @Linchenn, and @qjia7)

___
*[`tfjs-backend-webgl/src/backend_webgl.ts` line 1316 at r24](https://reviewable.io/reviews/tensorflow/tfjs/6853#-NDUubtN-tZnva-_fVrJ:-NDUubtN-tZnva-_fVrK:b-xdy54l) ([raw file](https://github.com/tensorflow/tfjs/blob/abb81a754d5e04bf25d618f659585895a5ae9982/tfjs-backend-webgl/src/backend_webgl.ts#L1316)):*
> ```XML
> 
>     if (!(channels in CHANNELS_TO_NUM_MAP)) {
>       throw new Error(`The color must be a non-empty subsequence of 'RGBA'.`);
> ```

could make the error message bit informative, 
`the channels must be a subsequence of 'RGBA', i.e. ('R', 'G', 'GB', 'RGB')



_Code quote:_
```XML
subsequence
```
___
*[`tfjs-backend-webgl/src/backend_webgl_test.ts` line 1123 at r24](https://reviewable.io/reviews/tensorflow/tfjs/6853#-NDV3liGAf0eRFJDYsc9:-NDV3liGAf0eRFJDYscA:b-9ibel7) ([raw file](https://github.com/tensorflow/tfjs/blob/abb81a754d5e04bf25d618f659585895a5ae9982/tfjs-backend-webgl/src/backend_webgl_test.ts#L1123)):*
> ```XML
> 
>     const gl = gpgpu.gl;
>     const texture = gl.createTexture();
> ```

Please verify texture created on another canvas can be shared with TFJS webgl backend's offscreen canvas?
___
*[`tfjs-backend-webgl/src/backend_webgl_test.ts` line 1307 at r24](https://reviewable.io/reviews/tensorflow/tfjs/6853#-NDUzJjL0iZUh7MUxefK:-NDUzJjL0iZUh7MUxefL:bskyp8e) ([raw file](https://github.com/tensorflow/tfjs/blob/abb81a754d5e04bf25d618f659585895a5ae9982/tfjs-backend-webgl/src/backend_webgl_test.ts#L1307)):*
> ```XML
> 
>     expect(a.shape).toEqual(logicalShape);
>     expect(a.dtype).toEqual('bool');
> ```

bool dtype might require value to be 0 or 1, here if texture value is uploaded to GPU as float32, it might create computational error.
We can throw error if user need to create 'bool' tensor or add the float to bool conversion logic to the shader.
___
*[`tfjs-backend-webgl/src/encode_matrix_gpu.ts` line 22 at r24](https://reviewable.io/reviews/tensorflow/tfjs/6853#-NDV8iG8BziuMS5jOK9M:-NDV8iG8BziuMS5jOK9N:b-uovbi) ([raw file](https://github.com/tensorflow/tfjs/blob/abb81a754d5e04bf25d618f659585895a5ae9982/tfjs-backend-webgl/src/encode_matrix_gpu.ts#L22)):*
> ```XML
> import * as shader_util from './shader_compiler_util';
> 
> export const CHANNELS_TO_NUM_MAP: Record < string, number >= {
> ```

offline discussion about this map for supported channels, it would be more flexible to support subset of the RGBA, which means the order should be settable by the user.

**Review by pyu10055 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/6853)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @lina128, @Linchenn, and @qjia7)

___
*[`tfjs-core/src/types.ts` line 174 at r26](https://reviewable.io/reviews/tensorflow/tfjs/6853#-NDYmJgU-jv75yc5V_po:-NDYmJgU-jv75yc5V_pp:bx6juev) ([raw file](https://github.com/tensorflow/tfjs/blob/988a27b5f4d155b2919d0b00b930f6fa83f12f2b/tfjs-core/src/types.ts#L174)):*
> ```XML
>   width: number;
>   // All non-empty permutation of all subset of 'RGBA'.
>   channels: 'A'|'B'|'G'|'R'|'AB'|'AG'|'AR'|'BA'|'BG'|'BR'|'GA'|'GB'|'GR'|'RA'|
> ```

It might be good to define this subset as a typescript type, so it can be reference it other places.

**Review by Linchenn — COMMENTED** [review]
Thank you all for reviewing and suggestions!

Now, the input texture supports all subset of 'RGBA' in any order (the order affects the orders of tensor values). 

Also, it should be noticed that, to use this API, users have to make sure the texture and TFJS are using same `gl`. If users want to create tensor from their custom canvas, they have to create a WebGL backend on this custom canvas.

> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/6853#-:-ND_Fpim1Uz_Jn9ktEd7:b8dspe5)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @lina128, @pyu10055, and @qjia7)

___
*[`tfjs-backend-webgl/src/backend_webgl.ts` line 192 at r8](https://reviewable.io/reviews/tensorflow/tfjs/6853#-NCgpR-PQ_F0Qnlerv15-r8-192:-ND_FhzOD9B47xvWOTS9:b8tdh0h) ([raw file](https://github.com/tensorflow/tfjs/blob/c60a220aca1a08b3e9ca028188234fa540b76d30/tfjs-backend-webgl/src/backend_webgl.ts#L192)):*
<details><summary><i>Previously, qjia7 (Jiajia Qin) wrote…</i></summary><blockquote>

So we require that the input texture is always an unpacked texture, right?
</blockquote></details>

Updated. Packed texture is allowed now.
___
*[`tfjs-backend-webgl/src/backend_webgl.ts` line 1316 at r24](https://reviewable.io/reviews/tensorflow/tfjs/6853#-NDUubtN-tZnva-_fVrJ:-NDWZNKlB3wsa90Q2lp9:b-896fix) ([raw file](https://github.com/tensorflow/tfjs/blob/abb81a754d5e04bf25d618f659585895a5ae9982/tfjs-backend-webgl/src/backend_webgl.ts#L1316)):*
<details><summary><i>Previously, pyu10055 (Ping Yu) wrote…</i></summary><blockquote>

could make the error message bit informative, 
`the channels must be a subsequence of 'RGBA', i.e. ('R', 'G', 'GB', 'RGB')
</blockquote></details>

Done.
___
*[`tfjs-backend-webgl/src/backend_webgl_test.ts` line 1123 at r24](https://reviewable.io/reviews/tensorflow/tfjs/6853#-NDV3liGAf0eRFJDYsc9:-ND_FThZ4_0tG-ONB4KD:b-4o6sf6) ([raw file](https://github.com/tensorflow/tfjs/blob/abb81a754d5e04bf25d618f659585895a5ae9982/tfjs-backend-webgl/src/backend_webgl_test.ts#L1123)):*
<details><summary><i>Previously, pyu10055 (Ping Yu) wrote…</i></summary><blockquote>

Please verify texture created on another canvas can be shared with TFJS webgl backend's offscreen canvas?
</blockquote></details>

You are right. As Kai said, there's no way to share textures between contexts. Updated doc for users.
___
*[`tfjs-backend-webgl/src/backend_webgl_test.ts` line 1307 at r24](https://reviewable.io/reviews/tensorflow/tfjs/6853#-NDUzJjL0iZUh7MUxefK:-NDWZQbt5l-9yuzQmGUS:b-896fix) ([raw file](https://github.com/tensorflow/tfjs/blob/abb81a754d5e04bf25d618f659585895a5ae9982/tfjs-backend-webgl/src/backend_webgl_test.ts#L1307)):*
<details><summary><i>Previously, pyu10055 (Ping Yu) wrote…</i></summary><blockquote>

bool dtype might require value to be 0 or 1, here if texture value is uploaded to GPU as float32, it might create computational error.
We can throw error if user need to create 'bool' tensor or add the float to bool conversion logic to the shader.
</blockquote></details>

Done.
___
*[`tfjs-backend-webgl/src/encode_matrix_gpu.ts` line 22 at r24](https://reviewable.io/reviews/tensorflow/tfjs/6853#-NDV8iG8BziuMS5jOK9M:-NDWZSj17ViruuMkBrRj:b-896fix) ([raw file](https://github.com/tensorflow/tfjs/blob/abb81a754d5e04bf25d618f659585895a5ae9982/tfjs-backend-webgl/src/encode_matrix_gpu.ts#L22)):*
<details><summary><i>Previously, pyu10055 (Ping Yu) wrote…</i></summary><blockquote>

offline discussion about this map for supported channels, it would be more flexible to support subset of the RGBA, which means the order should be settable by the user.
</blockquote></details>

Done.
___
*[`tfjs-core/src/types.ts` line 174 at r26](https://reviewable.io/reviews/tensorflow/tfjs/6853#-NDYmJgU-jv75yc5V_po:-ND_FeUx47-_sNdHldoD:b-896fix) ([raw file](https://github.com/tensorflow/tfjs/blob/988a27b5f4d155b2919d0b00b930f6fa83f12f2b/tfjs-core/src/types.ts#L174)):*
<details><summary><i>Previously, pyu10055 (Ping Yu) wrote…</i></summary><blockquote>

It might be good to define this subset as a typescript type, so it can be reference it other places.
</blockquote></details>

Done.

**Review by Linchenn — COMMENTED** [review]
using `backend.gpgpu.gl.isTexture(texture)` to validate texture and make sure the texture and tfjs are using same `gl`, otherwise WebGL just warns and returns wrong values.

> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/6853#-:-ND_KEKdAkDXRNhv-ZeD:bvmv966)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @lina128, @pyu10055, and @qjia7)

**Review by pyu10055 — APPROVED** [review]
great work, thanks!

> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/6853#-:-NDdCIoV9cNCuJ3lgjT6:b6rtb8b)* status: :shipit: complete! 2 of 1 approvals obtained (waiting on @lina128, @Linchenn, @pyu10055, and @qjia7)

___
*[`tfjs-backend-webgl/src/backend_webgl.ts` line 1320 at r28](https://reviewable.io/reviews/tensorflow/tfjs/6853#-NDd9z_z3ihmo3JkY1ND:-NDd9z_z3ihmo3JkY1NE:bwp108w) ([raw file](https://github.com/tensorflow/tfjs/blob/cad46bb0af838d1ec809a1f4051640d9ffc587d7/tfjs-backend-webgl/src/backend_webgl.ts#L1320)):*
> ```XML
>       throw new Error(
>           `The texture is invalid. Also, please make sure the texture and ` +
>           `the TFJS WebGL backend are using same canvas. If you want to use ` +
> ```

using the same cavas

**Review by Linchenn — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/6853)* status: :shipit: complete! 2 of 1 approvals obtained (waiting on @lina128, @pyu10055, and @qjia7)

___
*[`tfjs-backend-webgl/src/backend_webgl.ts` line 1320 at r28](https://reviewable.io/reviews/tensorflow/tfjs/6853#-NDd9z_z3ihmo3JkY1ND:-NDdD7XW8GjOnKMHehZd:boyexb3) ([raw file](https://github.com/tensorflow/tfjs/blob/a08123eecfaa1ef003bdec127a22a4423aa4350a/tfjs-backend-webgl/src/backend_webgl.ts#L1320)):*
<details><summary><i>Previously, pyu10055 (Ping Yu) wrote…</i></summary><blockquote>

using the same cavas
</blockquote></details>

Done. Thanks!

**Review by lina128 — APPROVED** [review]

**Review by pyu10055 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/6853)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @lina128, @Linchenn, @pyu10055, and @qjia7)

___
*[`tfjs-core/src/ops/tensor.ts` line 89 at r29](https://reviewable.io/reviews/tensorflow/tfjs/6853#-NDdsD9b5pFWqEfViuPz:-NDdsD9b5pFWqEfViuQ-:br11f8d) ([raw file](https://github.com/tensorflow/tfjs/blob/59d06982e0a99d1a968e5edb2136883e97af7d33/tfjs-core/src/ops/tensor.ts#L89)):*
> ```XML
>  * // channels of Pixel1...
>  *
>  * // For postprocessing on the GPU, it's possible to retrieve the texture
> ```

probably they should use the dataToGPU API for post processing, not grabbing the texture directly (otherwise texture is in packed format)

**Review by Linchenn — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/6853)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @lina128, @pyu10055, and @qjia7)

___
*[`tfjs-core/src/ops/tensor.ts` line 89 at r29](https://reviewable.io/reviews/tensorflow/tfjs/6853#-NDdsD9b5pFWqEfViuPz:-NDdw5gp6mKl5CrTgXsu:bqxagly) ([raw file](https://github.com/tensorflow/tfjs/blob/59d06982e0a99d1a968e5edb2136883e97af7d33/tfjs-core/src/ops/tensor.ts#L89)):*
<details><summary><i>Previously, pyu10055 (Ping Yu) wrote…</i></summary><blockquote>

probably they should use the dataToGPU API for post processing, not grabbing the texture directly (otherwise texture is in packed format)
</blockquote></details>

Great catch! Thanks!

**Review by pyu10055 — APPROVED** [review]
> Reviewed 1 of 7 files at r10, 1 of 1 files at r13, 2 of 3 files at r16, 5 of 9 files at r22, 1 of 5 files at r25, 1 of 1 files at r26, 1 of 3 files at r27, 1 of 2 files at r28, 1 of 1 files at r29, 1 of 1 files at r30, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/6853)* status: :shipit: complete! 2 of 1 approvals obtained (waiting on @lina128, @pyu10055, and @qjia7)

File: tfjs-backend-webgl/src/backend_webgl.ts
Comment (qjia7) [inline tfjs-backend-webgl/src/backend_webgl.ts:183]: nit: `color` -> `channels` to keep consistent?

File: tfjs-backend-webgl/src/backend_webgl.ts
Comment (qjia7) [inline tfjs-backend-webgl/src/backend_webgl.ts:199]: Do we need to provide a special path for `RGBA` by calling `EncodeMatrixPackedProgram`? In this case, we can avoid once `runWebGLProgram` to pack the tensor (`PackProgram`) in many situations.

**Review by qjia7 — COMMENTED** [review]
Thanks Lin. It looks great. Just some small comments.

File: tfjs-backend-webgl/src/backend_webgl.ts
Comment (Linchenn) [reply tfjs-backend-webgl/src/backend_webgl.ts:199]: RGBA input's packing schema is dense packing, while our tfjs's packing schema is square packing. As a result, even though both are packed (using all RGBA channels), we may need to transform the values through `EncodeMatrixPackedProgram`.

File: tfjs-backend-webgl/src/backend_webgl.ts
Comment (Linchenn) [reply tfjs-backend-webgl/src/backend_webgl.ts:183]: Done.

File: tfjs-core/src/tensor_util_env.ts
Comment (Linchenn) [reply tfjs-core/src/tensor_util_env.ts:32]: Great catch, done.

File: tfjs-core/src/ops/tensor.ts
Comment (Linchenn) [reply tfjs-core/src/ops/tensor.ts:98]: Thank you for the reminder and the reference!

File: tfjs-core/src/tensor_util_env.ts
Comment (Linchenn) [reply tfjs-core/src/tensor_util_env.ts]: Also, since we need to fill 'RGBA' as default if `channels` parameter is null. I moves this 'filling default' logics from backend to core, because to do this inferring shape in core, we also need to know the channels in core.

**Review by qjia7 — APPROVED** [review]
LGTM after bots become green. Thanks.

PS: Please also update your PR description to align with your current changes.

**Comment by qjia7** [comment]
cc @axinging You can continue the implementation in webgpu. Please refer to the doc of [Life of a Tensor](https://docs.google.com/document/d/1gRh63dNHcqUoSPYa85Xv_G5gG61Mfs2gZbkopO8pRoA/edit#heading=h.hxm6sghaxrgj) and the implementation here. Thanks.

**Pull request merged**
